[Skia] Fix GraphicsContextSkia::setLineDash when odd number of elements is provided

Scony opened the pull request:
#### 6d7d6a2d4ee87c3942ae1088f7412658fa8866d8
<pre>
[Skia] Fix GraphicsContextSkia::setLineDash when odd number of elements is provided
<a href="https://bugs.webkit.org/show_bug.cgi?id=271760">https://bugs.webkit.org/show_bug.cgi?id=271760</a>

Reviewed by Carlos Garcia Campos.

The spec states that if an odd number of elements is provided within dash array,
the array should be repeated to yield an even number of elements.
Therefore e.g. 1,2,3 should become 1,2,3,1,2,3.
* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
(WebCore::GraphicsContextSkia::setLineDash):

Canonical link: <a href="https://commits.webkit.org/276994@main">https://commits.webkit.org/276994@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f183be3a62b408bc8df61dfd8316900aaaad8771

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/45526 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/24652 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/48065 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/48194 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/41535 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/47833 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/28906 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/22046 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/37316 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/46104 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/21740 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/39285 "Passed tests") | [![loading-orange](https://github-production-user-asset-6210df.s3.amazonaws.com/3098702/291015173-08c448be-ac0a-4fd6-92a3-8165057445b7.png) 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18437 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; 1 api test failed or timed out; 1 api test failed or timed out; Running compile-webkit-without-change") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/19138 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/40370 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/3572 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/41849 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/40697 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/49931 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/20514 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/17056 "Passed tests") | [![loading-orange](https://github-production-user-asset-6210df.s3.amazonaws.com/3098702/291015173-08c448be-ac0a-4fd6-92a3-8165057445b7.png) 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/44386 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running run-layout-tests-in-stress-mode; 91 flakes 70 failures 1 missing results; Uploaded test results; 7 flakes 38 failures 1 missing results; Running compile-webkit-without-change") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/21820 "Built successfully") | | [![loading-orange](https://github-production-user-asset-6210df.s3.amazonaws.com/3098702/291015173-08c448be-ac0a-4fd6-92a3-8165057445b7.png) 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/43220 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; 3 api tests failed or timed out; 1 api test failed or timed out; Compiled WebKit; Running run-api-tests-without-change") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/22180 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6469 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/21508 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -786,7 +786,13 @@ void GraphicsContextSkia::setLineDash(const DashArray& dashArray, float dashOffs
+        repeatedDashArray.appendVector(dashArray);
+        m_skiaState.m_stroke.dash = SkDashPathEffect::Make(repeatedDashArray.data(), repeatedDashArray.size(), dashOffset);
+    } else
+        m_skiaState.m_stroke.dash = SkDashPathEffect::Make(dashArray.data(), dashArray.size(), dashOffset);
```
Looks good, I've noticed that in cairo (and chromium does the same) we also check that the array is not all 0 (for skia we only check the array is not empty) and set the dash to nullptr in that case too. We can do that in a follow up patch.

Scony replied:
Good point - skia doesn't validate all-0 case.

spenap replied:
@Scony pointed me to this as a low-hanging fruit I could look at, while I get more familiar with Skia in particular and rendering in general. 

While the initial idea was simple (something along the lines of `std::ranges::all_of(dashArray, isArrayAllZeros)`, I then started to look at what the other implementations do (as suggested by @carlosgcampos )

- Cairo indeed does that check -- and uses `all_of`
- Chromium, however, [doesn't](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/modules/canvas/canvas2d/base_rendering_context_2d.cc;l=1235;bpv=1;bpt=1). They verify that numbers are positive & finite

What Skia does, afterwards (in `SkDashPath::ValidDashPath` is ensure that all dashes are non-negative, and in the end, check that `length` is strictly positive and finite -- meaning it would allow `SkDashPathEffect::Make` to return `nullptr` when having received an array of zeros.

So my question here is now...

- do we still go ahead with a minor check for all-zero arrays, and continue relying on `ValidDashPath` for positive-finite checks?
- do we bring forwards all checks, and do them in one pass of the loop (i.e. check that no negative values are found, that they are all finite, and that we're not presented with an array of zeros)?

(It's probably a lot of thought for a possibly minor thing, but it I think it helps me understand how we all think, so I can align with you)

Scony committed:
[Skia] Fix GraphicsContextSkia::setLineDash when odd number of elements is provided
https://bugs.webkit.org/show_bug.cgi?id=271760

Reviewed by Carlos Garcia Campos.

The spec states that if an odd number of elements is provided within dash array,
the array should be repeated to yield an even number of elements.
Therefore e.g. 1,2,3 should become 1,2,3,1,2,3.
* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
(WebCore::GraphicsContextSkia::setLineDash):

Canonical link: https://commits.webkit.org/276994@main
